### ISSUE TITLE
Got many "EOF" errors after upgrade to 1.6.4

### ISSUE THREAD

**Issue by quexer** [aerospike/aerospike-client-go#75]
- we got 6000+ "EOF" errors every day after upgrade to 1.6.3 
- when we rollback to 1.6.2, the number reduce to 100
- we use a 3-nodes aerospike cluster in LAN, and the LAN is high speed and very stable
- any advice ?

thanks!

**Comment by GeertJohan** [comment]
Are you sure you are using 1.6.3, not 1.6.4?
Because 1.6.4 adds some error checking code that could display more EOF's. See commit 0f1e4cd.
They aren't occuring more often aince that commit, they're just handled better.

Do you have more information than just "EOF" ?

**Comment by khaf** [comment]
What command causes the EOF errors? (Get, Put, Query, ...)

**Comment by quexer** [comment]
@GeertJohan we check if the err  after every aerospike operation. if it isn't nil,  print the err.  then we go many "EOF" error in log file
@khaf  both "GET", "PUT", "QUERY" cause the EOF errors

**Comment by quexer** [comment]
@GeertJohan sorry I made a mistake.
Yes, we're using 1.6.4 and it show more "EOF" errors than 1.6.3.

any update about this issue?

**Comment by khaf** [comment]
This is expected if you are getting those EOF errors from queries. The old code was ignoring these errors, hence you weren't getting them.

Are these EOF errors impacting your operations? I would guess that increasing the timeout for queries would help a bit. 

Any guess regarding the percentage of bad requests to total?

**Comment by quexer** [comment]
There EOF errors are impacting my operation, although the percentage is not very high ( about 0.01 %)

How can I fix this ? Should I add retry logic to my application ?  
Can "GET" operation get benefit from  increasing the timeout for queries   ?

thanks！

**Comment by khaf** [comment]
I have an idea of what may be the problem. Let me work on it a bit and I'll come back to you.

Retries are baked in the client, and if you look at [`BasePolicy`](https://github.com/aerospike/aerospike-client-go/blob/master/policy.go), you can use `MaxRetries` and `SleepBetweenRetries` so that the client does the retries for you. `BasePolicy` is embedded in all policy objects.

**Comment by quexer** [comment]
I've dug into the code. When we do any operation with nil policy, the default policy will be used. 
It looks the default value "MaxRetries=2" doesn't help

**Comment by quexer** [comment]
@khaf any update ?

**Comment by khaf** [comment]
@quexer Sorry I was in a long vacation. I haven't been able to reproduce this problem, otherwise I would have had a fix by now. I also fail to see how the newer changes could have caused this.

Do you think you can provide a working snippet that shows this issue?

In all likelihood, you are just experiencing network problems.

**Comment by hectorj** [comment]
Hi!
I'm having what I believe is the same issue, and I tracked it down to: https://github.com/aerospike/aerospike-client-go/blob/master/write_command.go#L60

The *net.TCPConn gives an EOF on read there. Might be network or aerospike-server issue, still investigating that.

But about retries: No retry will be attempted with that error, because it goes through there: https://github.com/aerospike/aerospike-client-go/blob/master/command.go#L1045-L1050

Is it the desired behavior? Personally I'd like the client to do the retries, as these EOF are kind of random.

**Comment by hectorj** [comment]
Hi!
I'm having what I believe is the same issue, and I tracked it down to: https://github.com/aerospike/aerospike-client-go/blob/master/write_command.go#L60

The *net.TCPConn gives an EOF on read there. Might be network or aerospike-server issue, still investigating that.

But about retries: No retry will be attempted with that error, because it goes through there: https://github.com/aerospike/aerospike-client-go/blob/master/command.go#L1067-L1080 (to retry, we'd need a `continue` there)

Is it the desired behavior? Personally I'd like the client to do the retries, as these EOF are kind of random.

**Comment by hectorj** [comment]
@quexer : we found the reason for the EOF, please see https://github.com/aerospike/aerospike-server/issues/156

TL;DR : your `IdleTimeout` client's setting should be inferior to your server's `proto-fd-idle-ms`

@khaf : I noticed the Java client will retry on this kind of error (as you can see there: https://github.com/aerospike/aerospike-client-java/blob/master/client/src/com/aerospike/client/command/SyncCommand.java#L89-L92).

I'll make a quick PR to do the same here.

**Comment by khaf** [comment]
Thanks for your investigation and PR, I'll include the fix in the next release.

**Comment by khaf** [comment]
This has been merged in the master. Thanks everyone.
